### ISSUE TITLE
Support for incremental scanning of files (only scan the files that were modified since last run)

### ISSUE THREAD

**Issue by royriojas** [eslint/eslint#2998]
Running eslint thru a large code base takes something about from 6 to 8 seconds (depending on the amount of code in the files and how bad the code is in the first place)

Using a cache layer that remembers that files were not modified since last time they were considered valid hugely reduces the amount of time that the process take, depending on the number of files modified.

On a regular basis for a given feature only a handful of files are modified so having a way to execute the tool only on those files might save some time.

I have written a [cli wrapper over eslint](https://www.npmjs.com/package/eslinter) that does that, but having that into the `eslint` module would be really nice: 

This is an example of the tool executed over one of the folders of the project

```bash

# first run will take as much time as normal
eslinter 'src/**/*.js' 

>> config: .eslintrc
>> cache: true, format: eslint-friendly-formatter
>> Total files: 255, files to process: 255
>> Eslint validation complete. No errors found
>> time it took:  6539ms

# second and subsequent execution will operate only on changed files
eslinter 'src/**/*.js' 

>> config: .eslintrc
>> cache: true, format: eslint-friendly-formatter
>> Total files: 255, files to process: 4
>> Eslint validation complete. No errors found
>> time it took:  292ms

```
So as you see actually know which files changed makes a difference in the time it takes

**Comment by nzakas** [comment]
Where did the cache get written to?

**Comment by royriojas** [comment]
to the .cache folder inside the [flat-cache](https://www.npmjs.com/package/flat-cache) module which is used by the [file-entry-cache](https://www.npmjs.com/package/file-entry-cache). Cache is destroy if the flag --no-use-cache is passed. I haven't find a better place for the storage yet. I was thinking to use tmp, but since we were ignoring the node_modules folder that was not an issue in my project.

**Comment by ilyavolodin** [comment]
I can't say I'm crazy about writing something to the user system without them specifically requesting it. `/var/tmp` (or rather `windows/temp`) will be a problem on Windows, due to requirement of elevated permissions.

**Comment by royriojas** [comment]
@ilyavolodin  That's why the same folder for the module made sense for me. I realized this functionality could be helpful in other modules as well that's why I factorized it. but I didn't want to put the burden on the users to ask where to store the cache. but it can be specified. Some people prefer to know where things are being written in their systems :)

**Comment by nzakas** [comment]
For people who check in their dependencies, having a file show up could be disconcerting. If a file is to be written, it needs to be obvious.

I could see maybe creating a `.eslintcache` file in the cwd, but we would need to make this behavior opt-in so people would have the chance to add it to their `.gitignore` file.

**Comment by gyandeeps** [comment]
@royriojas Would you be interested in investigating this further?

**Comment by royriojas** [comment]
@gyandeeps sure! 

Using the `file-entry-cache` module is pretty straightforward to implement a functionality like this, we just need to define:

- the location of the cache
- a way to make it obvious
- a way for the cache to work when `eslint` is executed globally
- a way to delete the cache, or recreate the cache
- the maximum file size of the cache to prevent filling up the disk space

Let me know how I can contribute and I'm happy to help

**Comment by nzakas** [comment]
Here's my strawman:

1. People must specify `--cache` to enable caching
1. The cache information is stored in the current working directory as `.eslintcache`
1. It can be deleted manually and also if you run ESLint without `--cache`

I'm not sure about maximum size - seems like it shouldn't get big enough to worry about, unless I'm missing something.

**Comment by royriojas** [comment]
@nzakas, cool. I will take a look into this today

Another question, should the cache be available when Eslint is used as a module instead of a CLI tool? or should it only be available when using the CLI?

**Comment by nzakas** [comment]
It should be available via `CLIEngine`, since that's where the functionality would be used.

**Comment by zallek** [comment]
Am I right thinking that this new feature could be used to lint legacy projects ?
If you initialise the cache in a way that every existing file is "valid". It would allow eslint to only lint newly modified files and so would allow to refactor piece by piece existing code base.

That would be great because eslint is so good but it's nearly impossible to use on legacy projects as the amount of modifications is too important.

**Comment by IanVS** [comment]
I don't think that's really the intent of this feature.  Files are not added to the cache until they make a clean lint run (no errors, no warnings).  

There are a few strategies for adding ESLint to legacy projects:

* Set all rules to warn, not error.  Then, fix rules and set them to error are you are able.
* After setting rules to warn, implement a `--max-warnings` so that additional warnings cause the build to fail (https://github.com/eslint/eslint/issues/2769).
* Use a tool like [ESLint Nibble](https://github.com/IanVS/eslint-nibble) to get an overview of what rules are causing problems, then focus on fixing one rule at a time.
* Use [Dryer](https://github.com/willklein/dryer) to analyze your code and create an `.eslintrc` file which not cause any errors (on the roadmap to implement in ESLint `2.0.0`).

**Comment by TheSavior** [comment]
Is the cache invalidated if the configuration used to lint the file was changed? I'd like to enable this for our project but want to make sure things are linted again without the cache automatically if we ever change our configuration.

**Comment by ilyavolodin** [comment]
No, right now cache is invalidated only when target files change. You can manually reset cache though.

**Comment by victor-homyakov** [comment]
`--cache` results on a large SPA project (~2700 files):
- first scan (empty cache): ~55 seconds
- second scan (nothing changed): ~25 seconds

I expected much more significant boost. Scan without any changes should finish almost instantly, as fast as tree listing could complete.

**Comment by royriojas** [comment]
@victor-homyakov that's true, results are supposed to be shown almost immediately. Can you share the command line or the way you're running eslint in your project? I'm very curious about that.

**Comment by victor-homyakov** [comment]
OK, that was not the latest ESLint version. After update to 2.8.0 numbers are slightly different:
- first scan: ~35 seconds
- second scan (no changes): ~16 seconds

Command line:
`eslint --cache scripts/tau >eslint.log`

**Comment by royriojas** [comment]
@victor-homyakov 

So just to confirm, a couple more of questions:

- The first pass didn't generate any warnings or errors, right?
- You obtained the same results when running the following command:
eslint --cache 'scripts/**/*.js' > eslint.log

right?

**Comment by victor-homyakov** [comment]
- The first pass did generate multiple warnings and errors
- Results for the first and following passes are the same

**Comment by royriojas** [comment]
Yeah, I guess that is expected, the cache only stores the info for the files that passed validation. If any file didn't pass that will be inspected again. Maybe that's not really that clear, but that is the way it currently is.

**Comment by victor-homyakov** [comment]
Oh yes now I understand. Thanks!

**Comment by dawsbot** [comment]
What is the status of this? Why not incorporate a tool like [conf](https://github.com/sindresorhus/conf) to store cache information in a cross-platform way outside of the currently executing repository?

**Comment by royriojas** [comment]
@dawsbot not sure I understand your question, but Eslint does support storing the cache in external locations. 

```
Caching:
  --cache                      Only check changed files - default: false
  --cache-file path::String    Path to the cache file. Deprecated: use --cache-location - default: .eslintcache
  --cache-location path::String  Path to the cache file or directory
```

**Comment by galvarez421** [comment]
Would there be any performance advantage to storing info for files that do _not_ pass validation in the cache? I think the documentation is not as clear as it could be, because it says that only changed files will be checked, but essentially it is treating files with errors or warnings as always changed when it comes to caching. When I first read through the documentation, I thought that information for all linted files would be stored in the cache and that future lint reports for those files would come straight from the cache file, thereby saving time by not having to actually run a lint process again on those files (errors, warnings, or not), but that is not the intent of this feature I suppose?

If the cache only stores info for files that pass validation, how come there are properties for errorCount, warningCount, fixableErrorCount, and fixableWarningCount? I mean, would those ever not be 0?

**Comment by IanVS** [comment]
That's an interesting point.  If the file hasn't changed, it seems like we should be able to use the results of the previous cached run, regardless if there were errors or not.  The only files that really _need_ to be linted are new files and files that have been changed.  I think the proposal should probably be its own issue though, as this is an old, closed issue.  Would you mind opening a new one and tagging me on it, @galvarez421?
